### Discription:
This notebook implements a Convolutional Neural Network (ConvNet) on the MNIST dataset as demonstrated in the book “Deep Learning with Python.” The code has been reproduced and executed here solely for learning and educational purposes. All original concepts and code logic belong to the book and its author, and this notebook serves only as a personal practice implementation.

## Building a simple convnet:

In [11]:
import os
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"  # 0 = hide all, 1 = hide INFO, 2 = hide WARNING, 3 = hide ERROR
#The above piece of code is to hide the warnings 

In [12]:
from tensorflow import keras
from tensorflow.keras import layers

inputs = keras.Input(shape=(28,28,1),name='input_layer') #since out images have length and width of 28 pixals and color channels=1 since the images in MNIST are in grayscale
x = layers.Conv2D(filters=32, kernel_size=3, activation='relu') (inputs)
x = layers.MaxPooling2D(pool_size=2) (x)
x = layers.Conv2D(filters=64, kernel_size=3, activation='relu') (x)
x = layers.MaxPooling2D(pool_size=2) (x)
x = layers.Conv2D(filters=128, kernel_size=3, activation='relu') (x)
x = layers.Flatten() (x)
outputs = layers.Dense(10, activation='softmax', name='output_layer') (x)

model = keras.Model(inputs=inputs, outputs=outputs, name='ConvNet_1')

In [13]:
#Summerizing the model
model.summary()

Model: "ConvNet_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 28, 28, 1)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_3 (Conv2D)               │ (None, 26, 26, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 13, 13, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_4 (Conv2D)               │ (None, 11, 11, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_3 (MaxPooling2D)  │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 3, 3, 128)      │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 1152)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ output_layer (Dense)            │ (None, 10)             │        11,530 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 104,202 (407.04 KB)

 Trainable params: 104,202 (407.04 KB)

 Non-trainable params: 0 (0.00 B)

*  Output of every Conv2D and MaxPooling2D layer is a rank-3 tensor of shape (height, width, channels).
*  The width and height dimensions tend to shrink as we go deeper in the model.
*  The first argument that is 'filters' controls the number of channels in the output of Conv2D layer.

After the last Conv2D layer we have an output tensor of shape (3,3,128) but we had to feed it to our output layer which is a Densly connected layer, these layers process inputs which are 1D while we had a 3D input for it, so we flatten the 3D outputs to 1D using a *Flatten* layer, before passing it to the dense layer.

Lastly a 10 way classification was done in teh output layer via a *Softmax* activation since we have 10 digit classes in the MNIST dataset.

## Compiling and training the model on MNIST:
The dataset consists of 60000 training samples and 10000 test samples

In [14]:
#Importing the dataset
from tensorflow.keras.datasets import mnist
(train_imgs,train_labels),(test_imgs,test_labels) = mnist.load_data()

#Scaling and reshaping the data
train_imgs = train_imgs.reshape(60000,28,28,1)
train_imgs = train_imgs.astype('float32') / 255

test_imgs = test_imgs.reshape(10000,28,28,1)
test_imgs = test_imgs.astype('float32') / 255

In [15]:
#compiling and training the model
model.compile(optimizer='rmsprop',
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy'])

model.fit(train_imgs, train_labels, epochs=5, batch_size=64) 

Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.8830 - loss: 0.3697
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.9857 - loss: 0.0463
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.9910 - loss: 0.0285
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.9930 - loss: 0.0226
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.9948 - loss: 0.0170


## Evalutating/testing the model on testdata:

In [16]:
test_loss, test_acc = model.evaluate(test_imgs, test_labels)
print(f"Test accuracy: {test_acc}")


313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.9902 - loss: 0.0361
Test accuracy: 0.9921000003814697


The model achived astonishing **99.21%** accuracy on the test data, it proves that convnets are much better than a model with only Densly connected layers at compute vision tasks.